web/source/kmwcallback.ts

```diff
@@ -414,40 +427,56 @@ class KeyboardInterface {
 
     var mismatch = false;
 
+    var assertNever = function(x: never): never {
+      // Could be accessed by improperly handwritten calls to `fullContextMatch`.
+      throw new Error("Unexpected object in fullContextMatch specification: " + x);
+    }
+
     // Stage two:  time to match against the rule specified.
     for(var i=0; i < rule.length; i++) {
-      if(typeof(rule[i]) == 'string') {
+      if(typeof rule[i] == 'string') {
         var str = rule[i] as string;
         if(str != context[i]) {
           mismatch = true;
           break;
         }
-      } else if(rule[i]['d'] !== undefined) {
-        var deadSpec = rule[i] as RuleDeadkey;
-        // We still need to set a flag here; 
-        if(deadSpec['d'] != context[i]) {
-          mismatch = true;
-        } else {
-          deadContext[i].set();
-        }
-      } else if(rule[i]['a'] !== undefined) {
-        var anySpec = rule[i] as ContextAny;
-        // TODO:  Remove the `string` requirement.
-        if(!this.any(i, context[i] as string, anySpec.a)) {
-          mismatch = true;
-        }
-      } else if(rule[i]['i'] !== undefined) {
-        var indexSpec = rule[i] as RuleIndex;
-        var ch = this._Index(indexSpec.i.s, indexSpec.i.o);
+      } else {
+        // TypeScript needs a cast to this intermediate type to do its discriminated union magic.
+        var r = rule[i] as ContextNonCharEntry;
+        switch(r.t) {
+          case 'd':
+            var deadSpec = r as RuleDeadkey;
```

Comment:
Do we need this because the discriminated union magic should be constraining the type of `r` to `RuleDeadkey` now, shouldn't it?